src/main/java/com/atlassian/db/replica/internal/ConnectionParameters.java

```diff
@@ -14,8 +12,13 @@
     private String catalog;
     private Map<String, Class<?>> typeMap;
     private Integer holdability;
+    private final Set<Connection> initializedConnections = Collections.newSetFromMap(new WeakHashMap<>());
 
     public void initialize(Connection connection) throws SQLException {
+        if (initializedConnections.contains(connection)) {
```

Comment:
What about changing isolation level for an already initialized connection ? Not sure if that is possible but reading through unit tests this come to my mind.

Reply:
I think it's the reason for another problem we see on prod. I'm working on it 😄 .

Reply:
Ok. This one was covered by setting the current connection state in another place, but `setReadOnly` was not and it was causing the problems. I refactored the code to set the properties in one place and to cover setReadOnly and other parameters.

Reply:
I put it in the same PR as it seems to be related.